header/metrics.go

```diff
@@ -0,0 +1,46 @@
+package header
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric/global"
+	"go.opentelemetry.io/otel/metric/instrument"
+	"go.opentelemetry.io/otel/metric/unit"
+)
+
+var meter = global.MeterProvider().Meter("header")
+
+// MonitorHead enables Otel metrics to monitor head.
+func MonitorHead(store Store) {
+	headC, _ := meter.AsyncInt64().Counter(
+		"head",
+		instrument.WithUnit(unit.Dimensionless),
+		instrument.WithDescription("Subjective head of the node"),
+	)
+
+	err := meter.RegisterCallback(
+		[]instrument.Asynchronous{
+			headC,
+		},
+		func(ctx context.Context) {
+			head, err := store.Head(ctx)
+			if err != nil {
+				headC.Observe(ctx, 0, attribute.String("err", err.Error()))
+				return
+			}
+
+			headC.Observe(
+				ctx,
+				head.Height,
+				attribute.String("hash", head.Hash().String()),
+				attribute.String("app_hash", head.AppHash.String()),
+				attribute.String("data_root", head.DAH.String()),
```

Comment:
Apologies for raising this concern late but I'm concerned these labels will spike our [metric label cardinality](https://grafana.com/blog/2022/02/15/what-are-cardinality-spikes-and-why-do-they-matter/). These three fields will be unique per block so we will be generating many series. 

Proposal to remove labels that would be unique per block (i.e. these three). `square_size` seems safe to preserve as a label because the # of possible values is constrained.

<img width="752" alt="Screen Shot 2022-07-26 at 4 32 27 PM" src="https://user-images.githubusercontent.com/3699047/181106721-41225c1b-98de-43f8-9a4a-cc1c395b7e45.png">

Reply:
Made a deep look into the question
* The cardinality issue seems related to the Prometheus inefficiencies and lack of scalability, and the solution that Prometheus provides is to use fewer series:
> When you instrument your code and add a new metric, sometimes you attach more context than you need. 
*
	* It's like if Ethereum would say: "you transfer more value than you need" instead of putting efforts into scaling. And how do they know that we attach more than we need? I believe this is a common response from huge well-established status quo overmanaged enterprise companies on performance issues with their tech. They don't wanna change their legacy code, as they are too lazy and clumsy.

* In our case, this data is sent per block every block time(~30 secs in Mamaki). Data spikes could happen if there is an influx of users/node operators. 
	* In case of the further incentivized testnet the number of participants fixed? So no spikes?
		*  We still don't know if we will continue using this setup for the mainnet.
	* Also, our spikes would be slightly different than the case described in the article, AFAIU. 
		* In their case, there is _one_ microservice sending series but has a spike of new users, while in our, _n_ nodes send the same data with the same sending data size and time patterns.

* The data we send here _is useful for local monitoring setup_. but is not useful and just duplicates for global network view.
	* Should we do early optimizations to split the code into two types of monitoring(one configuration for local and another for global view)? 
		* We still don't know if such spikes will be a real issue according to our unique patterns, so I don't think we have to spend time splitting these and forcing devs to think about what should be sent to global view or not. 
		* If performance becomes an issue, we can consider using a more scalable drop-in replacement for Prometheus, like VictoriaMetrics(should we do it from the beginning?)
		* We can also investigate if Solana had any performance issues with Prometheus, as their metrics patterns might be similar to ours.

Reply:
I expect other monitoring databases may have better performance with regards to high cardinality data but they still advocate avoiding it:

> CAUTION: Remember that every unique combination of key-value label pairs represents a new time series, which can dramatically increase the amount of data stored. Do not use labels to store dimensions with high cardinality (many different label values), such as user IDs, email addresses, or other unbounded sets of values.

- https://prometheus.io/docs/practices/naming/#labels

> High cardinality usually means a high number of [active time series](https://docs.victoriametrics.com/FAQ.html#what-is-an-active-time-series). High cardinality may lead to high memory usage and/or to a high percentage of [slow inserts](https://docs.victoriametrics.com/FAQ.html#what-is-a-slow-insert). The source of high cardinality is usually a label with a large number of unique values, which presents a big share of the ingested time series. The solution is to identify and remove the source of high cardinality with the help of [cardinality explorer](https://docs.victoriametrics.com/#cardinality-explorer).
- https://docs.victoriametrics.com/FAQ.html#what-is-high-cardinality

Are you proposing using an alternative to Prometheus or VictoriaMetrics? If so, I think this should be evaluated in one of our ADRs.

Reply:
For clarification: my proposal is to avoid high cardinality data by removing these labels.

If you feel strongly that we should preserve these labels then we can continue down the path of exploring alternative DBs that better support high cardinality data. I can explore other databases in ADR-010 because it already considers Prometheus and InfluxDB

Reply:
My main point above is that we need these values for the local setup. The primary use cases I assumed for that are local debugging and the ability for a node operator to look up some data related to the height quickly. Later on, I realized that the former should be done with tracing instead, and the latter with some form of network explorer/indexer specialized in storing all the data about blockchain. Putting all that pressure on metrics is unnecessary, so let's remove those. 

We can still consider using other solutions to Prometheus, but I don't feel strongly about that. If you see the need and potential overload of the network, then let's add this to ADR; if not, let's not come back to the question until there is a performance issue.

Reply:
Removed

Reply:
> The primary use cases I assumed for that are local debugging and the ability for a node operator to look up some data related to the height quickly.

One thought is that celestia-node still emits logs that contain block height and hash, ex.
```
2022-07-28T23:28:54.109Z	INFO	das	das/daser.go:282	sampled header	{"height": 153029, "hash": "7757A6368C9BE0833B644A1873DAD68FD93DA98BFFFE1C86FD1BC821CF3842AF", "square width": 16, "finished (s)": 0.482185124}
```

so for debugging locally, an operator or celestia-node developer may be able to get comparable information from logs rather than metrics. Defer to you on how best to capture traces or logs with this data.

Metric labels with a large # of possible values seems like a non-obvious foot gun that we should try to avoid. I'm open to suggestions on how to enforce this in the codebase. We can monitor cardinality (see [docs](https://grafana.com/docs/grafana-cloud/data-configuration/metrics/metrics-control-usage/control-prometheus-metrics-usage/cardinality-management/?pg=blog&plcmt=body-txt)) however this seems less helpful than catching and preventing implementation of metrics with high cardinality.

Reply:
As long I am working on celestia-node, I will take care of this and other devs who joined and learned this will follow along.